src/frames.ts

```diff
@@ -386,39 +402,53 @@ export class Frame {
       }
       url = helper.completeUserURL(url);
 
-      const frameTask = new FrameTask(this, progress);
-      const sameDocumentPromise = frameTask.waitForSameDocumentNavigation();
-      const navigateResult = await this._page._delegate.navigateFrame(this, url, referer).catch(e => {
-        // Do not leave sameDocumentPromise unhandled.
-        sameDocumentPromise.catch(e => {});
-        throw e;
-      });
-      let request: network.Request | undefined;
+      const sameDocument = helper.waitForEvent(progress, this, kNavigationEvent, (event: NavigationEvent) => !event.documentInfo);
+      const navigateResult = await this._page._delegate.navigateFrame(this, url, referer);
+
+      let event: NavigationEvent;
       if (navigateResult.newDocumentId) {
-        // Do not leave sameDocumentPromise unhandled.
-        sameDocumentPromise.catch(e => {});
-        request = await frameTask.waitForSpecificDocument(navigateResult.newDocumentId);
+        sameDocument.dispose();
```

Comment:
It looks like we could benefit from 
`DisposablePromise.race([dp1, dp2])` where all promises are disposed upon race resolution.